python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,432 @@
+            # Make sure that the invoker is bound to the main thread
+            self.moveToThread(QtGui.QApplication.instance().thread())
+
+        def __call__(self, fn, *args, **kwargs):
+            """
+            Asks the MainTheadInvoker to call a function with the provided parameters in the main
+            thread.
+            :param fn: Function to call in the main thread.
+            :param args: Array of arguments for the method.
+            :param kwargs: Dictionary of named arguments for the method.
+            :returns: The result from the function.
+            """
+            self._fn = lambda: fn(*args, **kwargs)
+            self._res = None
+
+            logger.debug("Sending ui request to main thread.")
+
+            QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
+
+            # If an exception has been thrown, rethrow it.
+            if self._exception:
+                raise self._exception
+            return self._res
+
+        @QtCore.Slot()
+        def _do_invoke(self):
+            """
+            Execute function and return result
+            """
+            try:
+                logger.debug("Invoking from main thread.")
+                self._res = self._fn()
+            except Exception, e:
+                self._exception = e
+
+    return MainThreadInvoker()
+
+
+class AuthenticationHandlerBase(object):
+    """
+    Base class for authentication requests. It handles locking reading cached credentials
+    on disk and writing newer credentials back. It also keeps track of any attempt to cancel
+    authentication.
+    """
+
+    def authenticate(self, host, login, http_proxy):
+        """
+        Does the actual authentication. Prompts the user and validates the credentials.
+        :param host Host to authenticate for.
+        :param login: User that needs authentication.
+        :param http_proxy: Proxy to connect to when authenticating.
+        :raises: TankAuthenticationError If the user cancels the authentication process,
+                 this exception will be thrown.
+        """
+        raise NotImplementedError
+
+    def _get_session_token(self, hostname, login, password, http_proxy):
+        """
+        Retrieves a session token for the given credentials.
+        :param hostname: The host to connect to.
+        :param login: The user to get a session for.
+        :param password: Password for the user.
+        :param http_proxy: Proxy to use. Can be None.
+        :returns: If the credentials were valid, returns a session token, otherwise returns None.
+        """
+        try:
+            return session_cache.generate_session_token(hostname, login, password, http_proxy)
+        except AuthenticationError:
+            return None
+
+    def raise_authentication_cancelled_error(self):
+        raise AuthenticationCancelled()
+
+
+class ConsoleAuthenticationHandlerBase(AuthenticationHandlerBase):
```

Comment:
While the AuthenticationHandleBase is pretty useless at this point (it used to do a lot more, but after all the refactoring it's a pretty pointless class I agree, so I removed it), I'd rather keep the inheritance tree for the console stuff. I quite like the fact that each class has a very specific role. Each class taken individually is easier to understand than if they were mashed together in my humble opinion. The only reason why the ui authentication class shares it's implementation with session renewal is because it's the login dialog class that actually changes behaviour. If I had time I would probably prefer to have two classes, one for the UI authentication and the other for UI session renewal.